src/electionguard/precompute_buffers.cpp

```diff
@@ -0,0 +1,243 @@
+        // between generations of two triples and a quad. By full it means
+        // we check how many quads are in the queue, to start with we will
+        // try 500 and see how that works. If the vendor wanted to pass the
+        // queue size in we could use that.
+        // for now we just go through the loop once
+        do {
+            // TODO - Can we tolerate not returning until we have finished
+            // generating two triples and a quadruple?
+            // If not we can get more elaborate with the populate_OK checking
+            getInstance().queue_lock.lock();
+            if (getInstance().populate_OK) {
+                unique_ptr<Triple> triple1;
+                unique_ptr<Triple> triple2;
+                unique_ptr<Quadruple> quad;
+
+                // generate two triples and a quadruple
+                triple1 = make_unique<Triple>(elgamalPublicKey);
+                TripleEntry *triple_entry1 = new TripleEntry();
+                triple_entry1->set(move(triple1));
+                getInstance().triple_queue.push(triple_entry1);
+               
+                triple2 = make_unique<Triple>(elgamalPublicKey);
+                TripleEntry *triple_entry2 = new TripleEntry();
+                triple_entry2->set(move(triple2));
+                getInstance().triple_queue.push(triple_entry2);
+
+                quad = make_unique<Quadruple>(elgamalPublicKey);
+                QuadrupleEntry *quad_entry = new QuadrupleEntry();
+                quad_entry->set(move(quad));
+                getInstance().quadruple_queue.push(quad_entry);
+                getInstance().queue_lock.unlock();
+            } else {
+                getInstance().queue_lock.unlock();
+                return;
+            }
+        } while (getInstance().quadruple_queue.size() < getInstance().max);
+    }
+
+    void PrecomputeBufferContext::stop_populate()
+    {
+        getInstance().queue_lock.lock();
+        getInstance().populate_OK = false;
+        getInstance().queue_lock.unlock();
+    }
+
+    uint32_t PrecomputeBufferContext::get_max_queue_size()
+    {
+        return getInstance().max;
+    }
+
+    uint32_t PrecomputeBufferContext::get_current_queue_size()
+    {
+        return getInstance().quadruple_queue.size();
+    }
+      
+    std::unique_ptr<TwoTriplesAndAQuadruple> PrecomputeBufferContext::getTwoTriplesAndAQuadruple()
+    {
+        unique_ptr<TwoTriplesAndAQuadruple> result = make_unique<TwoTriplesAndAQuadruple>();
+
+        // take a lock while we get the triples and a quadruple
+        getInstance().queue_lock.lock();
+
+        // make sure there are enough in the queues
+        if ((getInstance().triple_queue.size() >= 2) && (getInstance().quadruple_queue.size() >= 1)) {
+            TripleEntry *triple_entry1 = getInstance().triple_queue.front();
+            unique_ptr<Triple> triple1 = triple_entry1->get_triple();
+            delete triple_entry1;
```

Comment:
curious why we need explicit resource initialization here. were the smart pointers not working for you? i've generally tried to keep any explicit heap allocations within the stl lifecycle management objects but of course there are always exceptions to every rule

Reply:
I was having difficulty with putting the smart pointers into the queues. That has been solved and the TripleEntry and QuadrupleEntry objects have been removed along with the news and frees.